core/storage/wal.rs

```diff
@@ -53,16 +53,34 @@ pub trait Wal {
         write_counter: Rc<RefCell<usize>>,
     ) -> Result<CheckpointStatus>;
     fn sync(&mut self) -> Result<CheckpointStatus>;
+    fn get_max_frame(&self) -> u64;
+    fn get_min_frame(&self) -> u64;
+}
+
+#[derive(Copy, Clone)]
+enum SyncState {
+    Start,
+    Wait,
+}
+
+struct OngoingCheckpoint {
+    page: PageRef,
+    state: CheckpointState,
+    min_frame: u64,
+    max_frame: u64,
+    current_page: u64,
 }
 
 pub struct WalFile {
     io: Arc<dyn crate::io::IO>,
+    buffer_pool: Rc<BufferPool>,
 
+    sync_state: RefCell<SyncState>,
     syncing: Rc<RefCell<bool>>,
     page_size: usize,
 
-    ongoing_checkpoint: HashSet<usize>,
     shared: Arc<RwLock<WalFileShared>>,
+    ongoing_checkpoint: OngoingCheckpoint,
```

Comment:
this would be better as an Option without the ::Start variant because there isn't always an ongoing checkpoint happening

Reply:
maybe in another pr